File: app/src/main/java/com/haishinkit/rtmp/messages/RTMPSetChunkSizeMessage.java
package com.haishinkit.rtmp.messages;

import java.nio.ByteBuffer;

import com.haishinkit.rtmp.RTMPConnection;
import com.haishinkit.rtmp.RTMPSocket;

public final class RTMPSetChunkSizeMessage extends RTMPMessage {
    private static final int CAPACITY = 4;

    private int size = 0;

    public RTMPSetChunkSizeMessage() {
        super(Type.CHUNK_SIZE);
    }

    public int getSize() {
        return size;
    }

    public RTMPSetChunkSizeMessage setSize(final int size) {
        this.size = size;
        return this;
    }

    @Override
    public ByteBuffer encode(final RTMPSocket socket) {
        if (socket == null) {
            throw new IllegalArgumentException();
        }
        ByteBuffer buffer = ByteBuffer.allocate(CAPACITY);
        buffer.putInt(getSize());
        return buffer;
    }

    @Override
    public RTMPMessage decode(final RTMPSocket socket, final ByteBuffer buffer) {
        return setSize(buffer.getInt());
    }

    @Override
    public RTMPMessage execute(final RTMPConnection connection) {
        connection.getSocket().setChunkSizeC(getSize());
        return this;
    }
}

File: haishinkit/src/main/java/com/haishinkit/rtmp/messages/RTMPSetChunkSizeMessage.kt
package com.haishinkit.rtmp.messages

import java.nio.ByteBuffer

import com.haishinkit.rtmp.RTMPConnection
import com.haishinkit.rtmp.RTMPSocket

/**
 * 5.4.1. Set Chunk Size (1)
 */
internal class RTMPSetChunkSizeMessage : RTMPMessage(RTMPMessage.Type.CHUNK_SIZE) {
    var size: Int = 0

    override fun encode(socket: RTMPSocket): ByteBuffer {
        val buffer = ByteBuffer.allocate(CAPACITY)
        buffer.putInt(size)
        return buffer
    }

    override fun decode(buffer: ByteBuffer): RTMPMessage {
        size = buffer.int
        return this
    }

    override fun execute(connection: RTMPConnection): RTMPMessage {
        connection.socket.chunkSizeC = size
        return this
    }

    companion object {
        private val CAPACITY = 4
    }
}
